crates/cranelift/src/func_environ.rs

```diff
@@ -865,14 +870,82 @@ impl<'module_environment> FuncEnvironment<'module_environment> {
+    fn ensure_table_exists(&mut self, func: &mut ir::Function, index: TableIndex) {
+        if !self.tables[index].is_reserved_value() {
+            return;
+        }
+
+        let pointer_type = self.pointer_type();
+
+        let (ptr, base_offset, current_elements_offset) = {
+            let vmctx = self.vmctx(func);
+            if let Some(def_index) = self.module.defined_table_index(index) {
+                let base_offset =
+                    i32::try_from(self.offsets.vmctx_vmtable_definition_base(def_index)).unwrap();
+                let current_elements_offset = i32::try_from(
+                    self.offsets
+                        .vmctx_vmtable_definition_current_elements(def_index),
+                )
+                .unwrap();
+                (vmctx, base_offset, current_elements_offset)
+            } else {
+                let from_offset = self.offsets.vmctx_vmtable_import_from(index);
+                let table = func.create_global_value(ir::GlobalValueData::Load {
+                    base: vmctx,
+                    offset: Offset32::new(i32::try_from(from_offset).unwrap()),
+                    global_type: pointer_type,
+                    flags: MemFlags::trusted().with_readonly(),
+                });
+                let base_offset = i32::from(self.offsets.vmtable_definition_base());
+                let current_elements_offset =
+                    i32::from(self.offsets.vmtable_definition_current_elements());
+                (table, base_offset, current_elements_offset)
+            }
+        };
+
+        let base_gv = func.create_global_value(ir::GlobalValueData::Load {
+            base: ptr,
+            offset: Offset32::new(base_offset),
+            global_type: pointer_type,
+            flags: MemFlags::trusted(),
+        });
+        let bound_gv = func.create_global_value(ir::GlobalValueData::Load {
+            base: ptr,
+            offset: Offset32::new(current_elements_offset),
+            global_type: ir::Type::int(
+                u16::from(self.offsets.size_of_vmtable_definition_current_elements()) * 8,
+            )
+            .unwrap(),
+            flags: MemFlags::trusted(),
+        });
+
+        let element_size = u64::from(
+            self.reference_type(self.module.table_plans[index].table.wasm_ty.heap_type)
+                .bytes(),
+        );
+
+        self.tables[index] = func.create_table(ir::TableData {
+            base_gv,
+            min_size: Uimm64::new(0),
+            bound_gv,
+            element_size: Uimm64::new(element_size),
+            index_type: I32,
+        });
+    }
+
     fn get_or_init_func_ref_table_elem(
         &mut self,
         builder: &mut FunctionBuilder,
         table_index: TableIndex,
-        table: ir::Table,
         index: ir::Value,
     ) -> ir::Value {
         let pointer_type = self.pointer_type();
+        self.ensure_table_exists(builder.func, table_index);
+        let table = self.tables[table_index];
```

Comment:
It seems like every time we call `ensure_table_exists` we then do `let table = self.tables[table_index];`

Might be worth returning that value from `ensure_table_exists` to clean up call sites a bit.

Reply:
I wish I could! In this PR it would be fine, since the values in `self.tables` are `Copy`, but later I'm going to have to extend it to a full struct that I'd want to return a borrow of. Unfortunately, if you have a `fn f(&'_ mut T) -> &'_ T`, the shared borrow that's returned acts like a mutable borrow, for reasons I don't entirely understand; the mutability is apparently attached to the lifetime. So I'll have to introduce this pattern later anyway.

I've also considered eagerly creating all the associated table data so that `ensure_table_exists` can go away, but that seemed like a larger change I didn't want to think through.